### PR TITLE
cfep-17: Handling pin backports and dependency rebuilds

### PR DESCRIPTION
**Pull request by CJ-Wright**
Thoughts @beckermr?

### CODE REVIEW

File: cfep-17.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+This CFEP proposes to automatically create branches for the pinned versions of pinned packages.
+
+This approach would have signifigant benefits:
+1. Make migrations easier for the bot, since the bot can issue PRs into the master and pinned branches
+2. Enable backporting of fixes for pins
+3. Provide the groundwork for backporting of updates to packages
+
+## Specification
+On merge of a PR into the pinned feedstock detect if the new version is beyond its stated compatibility, if so create a branch.
+The bot will only issue migration PRs into master and branch versions which are listed in the pinnings or a migration.
```
Comment (beckermr) [inline cfep-17.md:25]: bot behavior doesn't belong here - this cfep is only about automatically making pinning branches

File: cfep-17.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+The branches will need a consistent naming, for instance for boost `1.70.x`.
```
Comment (beckermr) [inline cfep-17.md:27]: i'd suggest something more unique and specified here in the cfep. For example, `cf-pinning-1.70.x` would help users know this is an automated branch.

File: cfep-17.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+The `x` acts as a filler for parts of the version number not provided by the `max_pin`.
+
+The bot will open version bump PRs matchign that version spec (for instnace, `1.70.1`) into that branch.
```
Comment (beckermr) [inline cfep-17.md:30]: again what the bot does is not relevant here

File: cfep-17.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+The CF global pinnings and the migrations that update the pinnings are important to maintaining consistancy across the CF ecosystem.
+However, the discrepency between the current pin and the version on the master (usually only) branch on the feedstock creates issues.
+
+For instance, if an update needs to be applied to a pinned package (potentially due to a depdendency migration) and the master branch has moved past the pin then a branch needs to be made for the pin.
+This CFEP proposes to automatically create branches for the pinned versions of pinned packages.
+
+This approach would have signifigant benefits:
+1. Make migrations easier for the bot, since the bot can issue PRs into the master and pinned branches
+2. Enable backporting of fixes for pins
+3. Provide the groundwork for backporting of updates to packages
```
Comment (beckermr) [inline cfep-17.md:21]: We should flesh this out more.

File: cfep-17.md (hunk 2 above)
Comment (isuruf) [reply cfep-17.md:27]: what about python 3.6 branch?

Comment (beckermr) [reply cfep-17.md:27]: What do you mean? Should we rename it according to this spec? Yes if we want automated updates.

**Comment by isuruf** [comment]
I find it confusing if the bot makes branches on feedstocks I maintain. Here's an alternative proposal,
- A bot makes an issue asking to make a branch
- Maintainer says yes by adding a label
- A github action will create the branch
makes it less confusing

**Comment by CJ-Wright** [comment]
The creation of the branch is an implementation detail, mostly. It is the way by which we can have multiple versions accessible to PR at once. I would say that the maintainer consent is provided by merging the PR that breaks the current pin. Having to respond to an issue seems like a bit of a burden for something most maintainers won't need to worry about.

**Comment by beckermr** [comment]
> The creation of the branch is an implementation detail, mostly.

Right. The issue here is one of consent. 

I am on the fence on this issue and am ok either way.

**Comment by isuruf** [comment]
Whatever you decide here, please don't make it the default for feedstocks that I maintain.

**Comment by beckermr** [comment]
> Whatever you decide here, please don't make it the default for feedstocks that I maintain.

Is opt-out too much to ask, or would you prefer opt-in?

**Comment by isuruf** [comment]
>Is opt-out too much to ask

I'm on 210 teams, so yeah.

**Comment by CJ-Wright** [comment]
Additionally, is the choice a real one? If a maintainer turns down this option, where does that put us, back where we are now? I don't think that is tenable.

**Comment by beckermr** [comment]
> Additionally, is the choice a real one? If a maintainer turns down this option, where does that put us, back where we are now? I don't think that is tenable.

This goes back to the discussions we've been having on keybase. How much power do we exert to  get migrations to go through?

**Comment by isuruf** [comment]
Let's say for example, you create a branch for clang 9, I won't be maintaining it because every time there's a rebuild, I have to build it on the ppc64le server and the aarch64 server. So I won't be merging any PRs unless you commit to doing the work yourself.

**Comment by beckermr** [comment]
Clang 9 is a very hefty example that updating via the bot is honestly not possible. 

A question @isuruf. Do you view people responding to an issue by adding a label as consent for the existence of the branch or as a commitment to maintain that branch too?

I think the difference here matters. We only need the branch to exist so that we can issue prs to it. We do not always need it to be maintained, though we may sometimes.

**Comment by isuruf** [comment]
As a commitment. If there's no branch, people will not try to send patches to these older versions.

File: cfep-17.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
+
+<table>
+<tr><td> Title </td><td> Branches for Globally Pinned Packages </td>
+<tr><td> Status </td><td> Draft </td></tr>
```
Comment (beckermr) [inline cfep-17.md]:
```suggestion
<tr><td> Status </td><td> Deferred </td></tr>
```

**Comment by beckermr** [comment]
merging as deferred

**Pull request merged**